Kafka-14420 Use incrementalAlterConfigs API for syncing topic configurations (KIP-894)

tinaselenge opened the pull request:
Implements https://cwiki.apache.org/confluence/display/KAFKA/KIP-894%3A+Use+incrementalAlterConfigs+API+for+syncing+topic+configurations

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tinaselenge committed:
KAFKA-14420: Fix bug in syncTopicConfigs function of mirrormaker2
Add a test that catches the bug

tinaselenge committed:
Use incrementalAlterConfigs() instead of the deprecated alterConfigs() when syncing topic configurations

tinaselenge committed:
Introduce a new settings for using incrementalAlterConfigs API
Extend ConfigPropertyFiler and DefaultConfigPropertyFilter with a new option to choose which cluster's default to use

tinaselenge committed:
Add unit tests for use.incremental.alter.configs setting

tinaselenge committed:
Merge branch 'trunk' into KAFKA-14420

tinaselenge committed:
Update after rebasing trunk

tinaselenge committed:
tidy up tests

tinaselenge committed:
Merge branch 'trunk' into KAFKA-14420

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -117,6 +122,16 @@ public MirrorSourceConnector() {
+        this.configPropertyFilter = configPropertyFilter;
+        this.useIncrementalAlterConfigs = useIncrementalAlterConfigs;
+        this.targetAdminClient = targetAdmin;
+        
```
We're missing a closing bracket here!

tinaselenge replied:
oops, that got lost when rebasing and resolving merge conflicts!

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -64,6 +67,7 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
```
This is already imported just above

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -539,11 +596,20 @@ Map<String, Config> describeTopicConfigs(Set<String> topics)
-            .filter(x -> shouldReplicateTopicConfigurationProperty(x.name()))
-            .collect(Collectors.toList());
+        List<ConfigEntry> entries;
+        if (useIncrementalAlterConfigs == MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIG) {
```
We should use `equals()` to compare String objects

mimaison replied:
Both branches are very similar, can we use merge them and only conditionally do the `isDefault()` check?

C0urante replied:
Do we need to do the `isDefault` check at all? I was under the impression that `ConfigPropertyFilter:: shouldReplicateSourceDefault` would be used to control replication of defaults regardless of whether we use the deprecated or incremental alter configs API.

tinaselenge replied:
I have merged them and included conditional isDefault() check in the filter. We basically want to always include default properties for incrementalAlterConfigs API, and then shouldReplicateSourceDefault should determine which operation should be used (SET/DELETE). For the deprecated API, shouldReplicateSourceDefault would be used to filter default properties. We can remove this filter, when we fully remove the deprecated API later.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/ConfigPropertyFilter.java:
```diff
@@ -27,6 +28,8 @@ public interface ConfigPropertyFilter extends Configurable, AutoCloseable {
 
     boolean shouldReplicateConfigProperty(String prop);
 
+    boolean shouldReplicateSourceDefault(ConfigEntry.ConfigSource source);
```
We need a default method here otherwise all existing implementations of this interface will break.

mimaison replied:
In the KIP the argument was the config name as a String, why is it a ConfigEntry.ConfigSource object here?

mimaison replied:
Can we also add some javadoc?

mimaison commented on clients/src/test/java/org/apache/kafka/clients/admin/AdminClientTestUtils.java:
```diff
@@ -78,6 +80,19 @@ public static ListTopicsResult listTopicsResult(String topic) {
+    public static AlterConfigsResult alterConfigsResult(ConfigResource cr, Throwable t) {
+        KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
+        Map<ConfigResource, KafkaFuture<Void>> futures = new HashMap<>();
+        futures.put(cr, future);
```
We can use singletonMap here:
```suggestion
Map<ConfigResource, KafkaFuture<Void>> futures = Collections.singletonMap(cr, future);
```

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/DefaultConfigPropertyFilter.java:
```diff
@@ -30,6 +31,9 @@ public class DefaultConfigPropertyFilter implements ConfigPropertyFilter {
     
     public static final String CONFIG_PROPERTIES_EXCLUDE_CONFIG = "config.properties.exclude";
     public static final String CONFIG_PROPERTIES_EXCLUDE_ALIAS_CONFIG = "config.properties.blacklist";
+    public static final String USE_DEFAULTS_FROM = "use.defaults.from";;
```
We can remove the trailing semi colon

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/DefaultConfigPropertyFilter.java:
```diff
@@ -30,6 +31,9 @@ public class DefaultConfigPropertyFilter implements ConfigPropertyFilter {
     public static final String CONFIG_PROPERTIES_EXCLUDE_CONFIG = "config.properties.exclude";
     public static final String CONFIG_PROPERTIES_EXCLUDE_ALIAS_CONFIG = "config.properties.blacklist";
+    public static final String USE_DEFAULTS_FROM = "use.defaults.from";;
+    private static final String USE_DEFAULTS_FROM_DOC = "Which cluster's default to use when syncing topic configurations.";
```
Should it be `defaults` instead of `default`?

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/DefaultConfigPropertyFilter.java:
```diff
@@ -40,11 +44,13 @@ public class DefaultConfigPropertyFilter implements ConfigPropertyFilter {
                                                                    + "unclean\\.leader\\.election\\.enable, "
                                                                    + "min\\.insync\\.replicas";
     private Pattern excludePattern = MirrorUtils.compilePatternList(CONFIG_PROPERTIES_EXCLUDE_DEFAULT);
+    private String useDefaultsFrom = USE_DEFAULTS_FROM_DEFAULT;
```
Do we need this field? I think we could remove it and do
```
"source".equals(config.useDefaultsFrom())
```
in `shouldReplicateSourceDefault()`

tinaselenge replied:
It's used in configure() to read the properties and config is not global variable. If I make config private variable, remove and ` private String useDefaultsFrom`, then try to access config.useFromDefaults from `shouldReplicateSourceDefault()`, it doesn't seem to work.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -73,6 +73,18 @@ public class MirrorSourceConfig extends MirrorConnectorConfig {
+    public static final String USE_INCREMENTAL_ALTER_CONFIG = "use.incremental.alter.configs";
+    private static final String USE_INCREMENTAL_ALTER_CONFIG_DOC = "Deprecated. Which API to use for syncing topic configs. " +
+            "The valid values are \"requested\", \"required\" and \"never\". " +
+            "By default, set to \"requested\", which means the IncrementalAlterConfigs is being used for syncing topic configurations " +
```
`IncrementalAlterConfigs` -> `IncrementalAlterConfigs API`

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -73,6 +73,18 @@ public class MirrorSourceConfig extends MirrorConnectorConfig {
+            "and if any request receives an error from an incompatible broker, it will fallback to using the deprecated AlterConfigs API. " +
+            "If explicitly set to \"required\", the IncrementalAlterConfigs is used without the fallback logic. " +
+            "If explicitly set to \"never\", the AlterConfig is always used." +
+            "This setting will be removed in Kafka 4.0, therefore users should ensure that target broker is at least 2.3.0";
```
I think we should mention that the behavior of `required` will be used from Kafka 4.0

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -162,6 +174,15 @@ Duration syncTopicConfigsInterval() {
 
+    String useIncrementalAlterConfigs() {
+        String prop = getString(USE_INCREMENTAL_ALTER_CONFIG);
+        if (prop.equals(NEVER_USE_INCREMENTAL_ALTER_CONFIG) || prop.equals(REQUIRE_INCREMENTAL_ALTER_CONFIG)) {
```
Do we need this if/else block? The config is defined with a default value so it should already return `USE_INCREMENTAL_ALTER_CONFIG_DEFAULT` is not set explicitly.

tinaselenge replied:
This was  in case user sets to a value other than 3 valid values. Do you think we don't need this?

mimaison replied:
We can set the valid values when defining the config. For example see https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java#L354-L355

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -264,6 +285,13 @@ Duration consumerPollTimeout() {
+                    USE_INCREMENTAL_ALTER_CONFIG_DEFAULT,
+                    ConfigDef.Importance.LOW,
+                    USE_INCREMENTAL_ALTER_CONFIG_DOC
+            )
```
nit: All other configs put the closing parenthesis on the last line. Let's use the same formatting.

mimaison commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -73,6 +73,18 @@ public class MirrorSourceConfig extends MirrorConnectorConfig {
+            "If explicitly set to \"required\", the IncrementalAlterConfigs is used without the fallback logic. " +
+            "If explicitly set to \"never\", the AlterConfig is always used." +
+            "This setting will be removed in Kafka 4.0, therefore users should ensure that target broker is at least 2.3.0";
+    public static final String USE_INCREMENTAL_ALTER_CONFIG_DEFAULT = "requested";
```
Should we rename this field to `REQUESTED_INCREMENTAL_ALTER_CONFIG`?

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -72,8 +76,22 @@
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
+import org.mockito.Mockito;
```
Unused import

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -209,11 +227,34 @@ public void testConfigPropertyFiltering() {
+    public void testConfigPropertyFilteringWithAlterConfigs() {
+        MirrorSourceConnector connector = new MirrorSourceConnector(new SourceAndTarget("source", "target"),
+                new DefaultReplicationPolicy(), MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIG, new DefaultConfigPropertyFilter(), null);
+        ArrayList<ConfigEntry> entries = new ArrayList<>();
```
Left side can be `List<ConfigEntry>`

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +302,54 @@ public void testNewTopicConfigs() throws Exception {
+    }
+
+    @Test
+    public void testIncrementalAlterConfigsRequired() throws Exception {
```
Can we add a test that causes `incrementalAlterConfig()` to emit a `OpType.DELETE` op?

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +302,54 @@ public void testNewTopicConfigs() throws Exception {
+        Config config = new Config(entries);
+        doReturn(Collections.singletonMap(topic, config)).when(connector).describeTopicConfigs(any());
+        doReturn(alterConfigsResult(new ConfigResource(ConfigResource.Type.TOPIC, topic), new UnsupportedVersionException("Unsupported API"))).when(admin).incrementalAlterConfigs(any());
+        doNothing().when(connector).alterConfigs(any());
```
Can we verify the arguments passed to `incrementalAlterConfigs()` and `alterConfigs()` in these tests?

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +302,54 @@ public void testNewTopicConfigs() throws Exception {
+        MirrorSourceConnector connector = spy(new MirrorSourceConnector(new SourceAndTarget("source", "target"),
+                new DefaultReplicationPolicy(), MirrorSourceConfig.USE_INCREMENTAL_ALTER_CONFIG_DEFAULT, new DefaultConfigPropertyFilter(), admin));
+        final String topic = "testtopic";
+        List<ConfigEntry> entries = new ArrayList<>();
+        entries.add(new ConfigEntry("name-1", "value-1"));
```
Can be simplified with:
```suggestion
        List<ConfigEntry> entries = Collections.singletonList(new ConfigEntry("name-1", "value-1"));
```

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -140,6 +147,143 @@ public void testReplication() throws Exception {
                 "New topic was not replicated to backup cluster.");
     }
 
+    @Test
```
Should this test be in `MirrorConnectorsIntegrationBaseTest` instead?

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -140,6 +147,143 @@ public void testReplication() throws Exception {
+        mm2Config = new MirrorMakerConfig(mm2Props);
+
+        waitUntilMirrorMakerIsRunning(backup, CONNECTOR_LIST, mm2Config, PRIMARY_CLUSTER_ALIAS, BACKUP_CLUSTER_ALIAS);
+        waitUntilMirrorMakerIsRunning(primary, Collections.singletonList(MirrorHeartbeatConnector.class), mm2Config, BACKUP_CLUSTER_ALIAS, PRIMARY_CLUSTER_ALIAS);
```
Do we need to start 2 instances of MirrorMaker for this test? A single instance would simplify this test a bit.

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -140,6 +147,143 @@ public void testReplication() throws Exception {
+        backup.kafka().incrementalAlterConfigs(configOps);
+
+        // sleep few seconds to have MM2 sync topic configuration
+        Thread.sleep(TimeUnit.SECONDS.toMillis(3));
```
Can we avoid using `sleep()` directly and instead use `waitForCondition()`?

mimaison commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -140,6 +147,143 @@ public void testReplication() throws Exception {
+        Collection<AlterConfigOp> ops = new ArrayList<>();
+        ops.add(new AlterConfigOp(new ConfigEntry("delete.retention.ms", "2000"), AlterConfigOp.OpType.SET));
+        ops.add(new AlterConfigOp(new ConfigEntry("retention.bytes", "2000"), AlterConfigOp.OpType.SET));
+        Map<ConfigResource, Collection<AlterConfigOp>> configOps = new HashMap<>(1);
+        configOps.put(configResource, ops);
```
We can use:
```suggestion
        Map<ConfigResource, Collection<AlterConfigOp>> configOps = Collections.singletonMap(configResource, ops);
```

mimaison reviewed (commented):
Thanks for the PR! I left a few questions and suggestions.

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -73,6 +73,18 @@ public class MirrorSourceConfig extends MirrorConnectorConfig {
+            "The valid values are \"requested\", \"required\" and \"never\". " +
+            "By default, set to \"requested\", which means the IncrementalAlterConfigs is being used for syncing topic configurations " +
+            "and if any request receives an error from an incompatible broker, it will fallback to using the deprecated AlterConfigs API. " +
+            "If explicitly set to \"required\", the IncrementalAlterConfigs is used without the fallback logic. " +
```
This differs from what the KIP described, which was:

> When explicitly set to "required", MirrorMaker will use the IncrementalAlterConfigs API for syncing topic configurations. If it receives an error from an incompatible broker, the MirrorMaker will report this to the user and fail the connector.

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +540,37 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
+        log.trace("Syncing configs for {} topics.", configOps.size());
+        targetAdminClient.incrementalAlterConfigs(configOps).values().forEach((k, v) -> v.whenComplete((x, e) -> {
+            if (e != null) {
+                if (useIncrementalAlterConfigs.equals(MirrorSourceConfig.USE_INCREMENTAL_ALTER_CONFIG_DEFAULT)
```
We also need to fail the connector if `use.incremental.alter.configs` is set to `required` and the error is an `UnsupportedVersionException`.

tinaselenge replied:
I have addressed this by adding `context.raiseError(new UnsupportedVersionException())` however didn't test this in the unit tests. I tried mocking the API to return `UnsupportedVersionException`, but not sure how to check the error from the connect framework in the test.

C0urante replied:
We can mock the connector's context and ensure that the correct method has been invoked:
```java
    @Test
    public void testIncrementalAlterConfigsRequiredButUnsupported() throws Exception {
        MockAdminClient admin = spy(new MockAdminClient());
        ConnectorContext connectorContext = mock(ConnectorContext.class);
        MirrorSourceConnector connector = spy(new MirrorSourceConnector(new SourceAndTarget("source", "target"),
                new DefaultReplicationPolicy(), MirrorSourceConfig.REQUIRE_INCREMENTAL_ALTER_CONFIG, new DefaultConfigPropertyFilter(), admin));
        connector.initialize(connectorContext);
        final String topic = "testtopic";
        List<ConfigEntry> entries = Collections.singletonList(new ConfigEntry("name-1", "value-1"));
        Config config = new Config(entries);
        doReturn(Collections.singletonMap(topic, config)).when(connector).describeTopicConfigs(any());
        doReturn(alterConfigsResult(new ConfigResource(ConfigResource.Type.TOPIC, topic), new UnsupportedVersionException("Unsupported API"))).when(admin).incrementalAlterConfigs(any());
        doNothing().when(connector).deprecatedAlterConfigs(any());

        connector.syncTopicConfigs();
        verify(connectorContext).raiseError(isA(UnsupportedVersionException.class));
    }
```

tinaselenge replied:
Ah thank you. I should have spotted the initialize() function!

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +540,37 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
+                        && e instanceof UnsupportedVersionException) {
+                    //Fallback logic
+                    log.warn("The target cluster {} is not compatible with IncrementalAlterConfigs API. Therefore using deprecated AlterConfigs API for syncing topic configurations", sourceAndTarget.target(), e);
+                    alterConfigs(topicConfigs);
```
We probably don't want to directly invoke `alterConfigs` here, since on the `scheduler` thread a call to that method may already be happening.

Also, we shouldn't be syncing `topicConfigs` since those may have grown stale in between when we computed them and when the config alter request completed.

I personally don't think it's necessary to invoke `alterConfigs` when we notice the failure at all, and we can instead just wait for the next scheduled topic config sync. However, if that's not good enough for some reason, we should probably use the `scheduler` for the immediate follow-up call to `alterConfigs`; e.g., something like `scheduler.execute(this::syncTopicConfigs, "syncing topic configs (failover)"`

tinaselenge replied:
I agree with this, and removed the alterConfigs call from here. Setting the `use.incremental.alter.configs` to `never` should be enough here so that next time we sync configs, it should use the deprecated API.

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -500,9 +517,18 @@ static Map<String, String> configToMap(Config config) {
+    // visible for testing
+    // use deprecated alterConfigs API for broker compatibility back to 0.11.0
+    @SuppressWarnings("deprecation")
+    void alterConfigs(Map<String, Config> topicConfigs) {
```
Do you think it might help to rename this to `deprecatedAlterConfigs`, to help contrast with `incrementalAlterConfigs` and make things clearer for people who have to read this code base for the first time in the future?

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -96,6 +100,7 @@ public class MirrorSourceConnector extends SourceConnector {
     private Admin sourceAdminClient;
     private Admin targetAdminClient;
     private Admin offsetSyncsAdminClient;
+    private String useIncrementalAlterConfigs;
```
Since the value of this field can be altered on a different thread from the one that checks it, I think we want to add the `volatile` keyword here.

C0urante reviewed (commented):
Thanks for the PR!

Haven't made a complete pass yet (probably not worth it until Mickael's comments are addressed), but wanted to leave a few thoughts that don't appear to conflict with any of the review comments left so far.

tinaselenge committed:
Address review comments

tinaselenge committed:
Address review comments and tidy up

tinaselenge committed:
Address remaining review comments

tinaselenge commented:
@mimaison @C0urante Thank you so much for reviewing the PR. I think I have addressed all the review comments, please let me know if I have missed anything or you have any further comments. 

The jenkins build has failed due to mostly unrelated tests, except one integration test for mirrormaker, `MirrorConnectorsIntegrationSSLTest > testReplication() FAILED`. However, this test passes on my local machine.

tinaselenge committed:
Merge branch 'trunk' into KAFKA-14420

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +326,78 @@ public void testNewTopicConfigs() throws Exception {
+
+        // the next time we sync topic configurations, expect to use the deprecated API
+        connector.syncTopicConfigs();
+        verify(connector, times(2)).syncTopicConfigs();
```
There's no need to verify these invocations since we make both of them in this test case ourselves.

(This applies to other test cases as well, such as `testIncrementalAlterConfigsRequired`.)

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +541,40 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
     }
 
+    // visible for testing
+    @SuppressWarnings("deprecation")
```
This part can be removed now (🎉)
```suggestion
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +541,40 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
+                    useIncrementalAlterConfigs = MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIG;
+                } else if (useIncrementalAlterConfigs.equals(MirrorSourceConfig.REQUIRE_INCREMENTAL_ALTER_CONFIG)
+                        && e instanceof UnsupportedVersionException) {
+                    context.raiseError(new UnsupportedVersionException("use.incremental.alter.configs was set to \"required\" however the target cluster is not compatible with IncrementalAlterConfigs API"));
```
Two small things:

1. We should log an ERROR level message when this fails (it will be redundant in most cases, but it could come in handy in case the call to `ConnectorContext::raiseError` fails)
2. We should include the original `UnsupportedVersionException` in case there's other useful information in that exception's stack trace and/or message

```suggestion
                    log.error("Failed to sync configs for topic {} on cluster {} with IncrementalAlterConfigs API", k.name(), sourceAndTarget.target(), e);
                    context.raiseError(new ConnectException(
                            "use.incremental.alter.configs was set to \"required\", but the target cluster '"
                                    + sourceAndTarget.target() + "' is not compatible with IncrementalAlterConfigs API",
                            e
                    ));
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -577,6 +641,9 @@ boolean shouldReplicateTopicConfigurationProperty(String property) {
 
+    boolean shouldReplicateSourceDefault(String property) {
+        return configPropertyFilter.shouldReplicateSourceDefault(property);
+    }
```
Nit: Add an extra line after method definitions
```suggestion
    }

```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -540,10 +601,13 @@ Map<String, Config> describeTopicConfigs(Set<String> topics)
-            .filter(x -> x.source() != ConfigEntry.ConfigSource.STATIC_BROKER_CONFIG)
-            .filter(x -> shouldReplicateTopicConfigurationProperty(x.name()))
-            .collect(Collectors.toList());
+                    .filter(x -> (!useIncrementalAlterConfigs.equals(MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIG)
```
Isn't it possible that `useIncrementalAlterConfigs` changes between when we read it here and when we read it in `updateTopicConfigs`?

C0urante replied:
Also, doesn't this result in a change in behavior for new topic creation (where we also use this method)?

C0urante replied:
I think we may want to alter this method to take a boolean `incremental` (rename as you like) parameter and use that instead of checking the value of `useIncrementalAlterConfigs`.

That way, when creating new topics, we can hardcode that parameter to `false`, and when syncing topics, we can check the value of `useIncrementalAlterConfigs` once in the beginning, and then use that single check to determine whether we're going to use the  incremental or deprecated API.

C0urante replied:
This is looking better, but not quite--we still aren't handling the possibility that `useIncrementalAlterConfigs` changes in the middle of an invocation of `syncTopicConfigs`. (Left a more detailed comment in the review.)

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -540,10 +601,13 @@ Map<String, Config> describeTopicConfigs(Set<String> topics)
+                            || (!x.isDefault() && useIncrementalAlterConfigs.equals(MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIG)))
+                    .filter(x -> !x.isReadOnly() && !x.isSensitive())
+                    .filter(x -> x.source() != ConfigEntry.ConfigSource.STATIC_BROKER_CONFIG)
+                    .filter(x -> shouldReplicateTopicConfigurationProperty(x.name()))
```
This double-checks default values, which I don't believe is correct (we should rely solely on `ConfigPropertyFilter::shouldReplicateSourceDefault` for those).

Perhaps all of the logic can be moved to the `shouldReplicateTopicConfigurationProperty` method, including determining whether the property is a default or not (and how to handle that accordingly)?

tinaselenge replied:
I'm not quite if I understand this comment. Would you mind elaborating on why you think shouldReplicateTopicConfigurationProperty is double checking the default values? My understanding from the KIP was that this function should determine whether to replicate the config regardless of what values(default or not) it has.

C0urante replied:
Hmm... I was imagining something slightly different:

For the `ConfigPropertyFilter` interface, the same property would never be tested against both methods (`shouldReplicateSourceDefault` and `shouldReplicateConfigProperty`). This lines up with the language in the KIP, which is that users "will be able to configure this method to return true in order to include source cluster's default configurations in the replication" (it mentions nothing about also having to return true from `shouldReplicateConfigProperty`). It also avoids some redundancy (you shouldn't have to change multiple methods to accommodate a single property).

tinaselenge replied:
I now understand what you are saying. I have attempted to address the suggestion, please let me if it doesn't look right.

C0urante replied:
Looks good, thanks!

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/ConfigPropertyFilter.java:
```diff
@@ -27,6 +27,13 @@ public interface ConfigPropertyFilter extends Configurable, AutoCloseable {
 
     boolean shouldReplicateConfigProperty(String prop);
```
Can we add a note to this method that it does not handle replication of default properties, and that that logic is instead controlled by `shouldReplicateSourceDefault`?

C0urante replied:
(Depends on conversation on a different thread)

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/DefaultConfigPropertyFilter.java:
```diff
@@ -60,6 +65,11 @@ public boolean shouldReplicateConfigProperty(String prop) {
 
+    @Override
+    public boolean shouldReplicateSourceDefault(String prop) {
+        return useDefaultsFrom.equals("source");
```
We should check against the `config.properties.exclude` pattern as well.

tinaselenge replied:
I'm not sure about this, because shouldReplicateConfigProperty does that already.

C0urante replied:
(Depends on discussion on a different thread but for now I still believe this part needs to be changed)

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/DefaultConfigPropertyFilter.java:
```diff
@@ -30,6 +30,9 @@ public class DefaultConfigPropertyFilter implements ConfigPropertyFilter {
+    private static final String USE_DEFAULTS_FROM_DEFAULT = "target";
 
     private static final String CONFIG_PROPERTIES_EXCLUDE_DOC = "List of topic configuration properties and/or regexes "
                                                                 + "that should not be replicated.";
```
Can we add a note that this will also apply to properties with default values if `use.defaults.from` is set to `source`?

C0urante replied:
This still needs to be done, not sure why it was marked resolved?

tinaselenge replied:
I had changed it `Which cluster's defaults to use when syncing topic configurations that have default values.` meaning that it only applies to properties with default values.  I have updated it to include the source part, what do you think?

C0urante replied:
I was commenting on the `CONFIG_PROPERTIES_EXCLUDE_DOC` docstring, not  `USE_DEFAULTS_FROM_DEFAULT`. But I think this depends on the outcome of the discussion on the other thread regarding how we want to use the `ConfigPropertyFilter`.

tinaselenge replied:
Ah sorry. I had misunderstood it!

C0urante replied:
No worries! I can see how the GitHub UI could be misleading here. Looks good now 👍

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConfig.java:
```diff
@@ -73,6 +75,19 @@ public class MirrorSourceConfig extends MirrorConnectorConfig {
+    @Deprecated
+    public static final String USE_INCREMENTAL_ALTER_CONFIG = "use.incremental.alter.configs";
+    private static final String USE_INCREMENTAL_ALTER_CONFIG_DOC = "Deprecated. Which API to use for syncing topic configs. " +
+            "The valid values are \"requested\", \"required\" and \"never\". " +
+            "By default, set to \"requested\", which means the IncrementalAlterConfigs API is being used for syncing topic configurations " +
+            "and if any request receives an error from an incompatible broker, it will fallback to using the deprecated AlterConfigs API. " +
+            "If explicitly set to \"required\", the IncrementalAlterConfigs API is used without the fallback logic and +" +
+            "if it receives an error from an incompatible broker, the connector will fail." +
+            "If explicitly set to \"never\", the AlterConfig is always used." +
+            "This setting will be removed and the behaviour of \"required\" will be used in Kafka 4.0, therefore users should ensure that target broker is at least 2.3.0";
```
Nit: it might be easier/cleaner to use single quotes around property names instead of double quotes, so that we don't have to use escape sequences.

```suggestion
            "The valid values are 'requested', 'required' and 'never'. " +
            "By default, set to 'requested', which means the IncrementalAlterConfigs API is being used for syncing topic configurations " +
            "and if any request receives an error from an incompatible broker, it will fallback to using the deprecated AlterConfigs API. " +
            "If explicitly set to 'required', the IncrementalAlterConfigs API is used without the fallback logic and +" +
            "if it receives an error from an incompatible broker, the connector will fail." +
            "If explicitly set to 'never', the AlterConfig is always used." +
            "This setting will be removed and the behaviour of 'required' will be used in Kafka 4.0, therefore users should ensure that target broker is at least 2.3.0";
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -96,6 +99,7 @@ public class MirrorSourceConnector extends SourceConnector {
     private Admin sourceAdminClient;
     private Admin targetAdminClient;
     private Admin offsetSyncsAdminClient;
+    private volatile String useIncrementalAlterConfigs;
```
I think it may be a bit simpler to replace this with a boolean value. Most of the time, we're only concerned about which API (incremental or deprecated) that we use. The only time the distinction between requested and required comes into play is if we see an `UnsupportedVersionException` when trying to use the incremental API, and at that point, we can just use `config::useIncrementalAlterConfigs` to check the value and then (if necessary) update the value of the `useIncrementalAlterConfigs` boolean instance field.

C0urante replied:
(This still needs to be addressed)

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -41,12 +41,12 @@
 
 /**
- * Tests MM2 replication and failover logic for {@link IdentityReplicationPolicy}.
+ * Tests MM2 replication and fail over logic for {@link IdentityReplicationPolicy}.
```
This is fine as-is; "failover" is definitely a word.

C0urante replied:
This still needs to be reverted; not sure why it was marked resolved?

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -108,16 +108,16 @@ public void testReplication() throws Exception {
                 backupClient, consumerGroupName, PRIMARY_CLUSTER_ALIAS, "test-topic-1");
 
-        // Failover consumer group to backup cluster.
+        // Fail-over consumer group to back up cluster.
```
Nit: this should actually be "fail over" (two words) to signify the verb, instead of "failover" (one word), which is the noun.

For the record, I'd be fine with leaving this as-is, but if we're going to change it, we should change it to the right thing.

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -108,16 +108,16 @@ public void testReplication() throws Exception {
             primaryConsumer.commitAsync();
 
-            assertTrue(primaryConsumer.position(new TopicPartition("test-topic-1", 0)) > 0, "Consumer failedover to zero offset.");
+            assertTrue(primaryConsumer.position(new TopicPartition("test-topic-1", 0)) > 0, "Consumer failed over to zero offset.");
```
This means the opposite of what it should.
```suggestion
            assertTrue(primaryConsumer.position(new TopicPartition("test-topic-1", 0)) > 0, "Consumer should have failed over to zero offset.");
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -151,13 +151,13 @@ public void testReplicationWithEmptyPartition() throws Exception {
         }
 
-        // one way replication from primary to backup
+        // one way replication from primary to back up
```
This is correct as-is; "backup" is the right word here.

C0urante reviewed (commented):
Thanks Tina! Left a few more comments.

I think for some of the unhandled cases here, it would also be nice to add testing coverage to ensure that things are working as expected now and that we don't regress in the future. FWIW, I still haven't gotten to look too much at the testing changes; I'm assuming that any logical errors here aren't covered or are covered incorrectly, since otherwise the tests would be failing.

tinaselenge committed:
Address further review comments

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/ConfigPropertyFilter.java:
```diff
@@ -25,8 +25,19 @@
 public interface ConfigPropertyFilter extends Configurable, AutoCloseable {
 
+    /**
+     * Specifies whether to replicate the given topic configuration whether it has a default value or not.
+     * The configurations with default values are controlled by shouldReplicateSourceDefault.
```
1. This still needs to be updated if we make `shouldReplicateConfigProperty` and `shouldReplicateSourceDefault` handle distinct properties
2. We can tidy up the Javadocs a bit by including a link to the relevant method

```suggestion
     * Specifies whether to replicate the given topic configuration whether it has a default value or not.
     * The configurations with default values are controlled by {@link #shouldReplicateSourceDefault(String)}. 
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/IdentityReplicationIntegrationTest.java:
```diff
@@ -41,12 +41,12 @@
  * Tests validate that records sent to the primary cluster arrive at the backup cluster. Then, a consumer group is
  * migrated from the primary cluster to the backup cluster. Tests validate that consumer offsets
- * are translated and replicated from the primary cluster to the backup cluster during this failover.
+ * are translated and replicated from the primary cluster to the backup cluster during this fail over.
```
This should also be reverted

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +543,41 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
+                if (useIncrementalAlterConfigs.equals(MirrorSourceConfig.REQUEST_INCREMENTAL_ALTER_CONFIG)
+                        && e instanceof UnsupportedVersionException) {
+                    //Fallback logic
+                    log.warn("The target cluster {} is not compatible with IncrementalAlterConfigs API. Therefore using deprecated AlterConfigs API for syncing topic configurations", sourceAndTarget.target(), e);
```
Nit: this line's pretty long, can we split it up?

```suggestion
                    log.warn(
                            "The target cluster {} is not compatible with IncrementalAlterConfigs API. "
                                    + "Therefore using deprecated AlterConfigs API for syncing topic configurations",
                            sourceAndTarget.target(),
                            e
                    );
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -365,11 +383,13 @@ void syncTopicAcls()
         Map<String, Config> targetConfigs = sourceConfigs.entrySet().stream()
-            .collect(Collectors.toMap(x -> formatRemoteTopic(x.getKey()), x -> targetConfig(x.getValue())));
+            .collect(Collectors.toMap(x -> formatRemoteTopic(x.getKey()), x -> targetConfig(x.getValue(), incremental)));
         updateTopicConfigs(targetConfigs);
```
We should also use `incremental` to determine which kind of topic alter API we use in `updateTopicConfigs`.

We can either do this by adding an `incremental` parameter to `updateTopicConfigs`, or by getting rid of that method entirely and just inlining it here.

C0urante reviewed (commented):
Thanks @tinaselenge, this is looking better. I think there are a few comments from the last round that went by unaddressed; I've resolved any that I think still need to be touched on.

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +543,41 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
+                    useIncrementalAlterConfigs = MirrorSourceConfig.NEVER_USE_INCREMENTAL_ALTER_CONFIG;
+                } else if (useIncrementalAlterConfigs.equals(MirrorSourceConfig.REQUIRE_INCREMENTAL_ALTER_CONFIG)
+                        && e instanceof UnsupportedVersionException) {
+                        log.error("Failed to sync configs for topic {} on cluster {} with IncrementalAlterConfigs API", k.name(), sourceAndTarget.target(), e);
+                        context.raiseError(new ConnectException("use.incremental.alter.configs was set to \"required\", but the target cluster '"
+                                + sourceAndTarget.target() + "' is not compatible with IncrementalAlterConfigs API", e));
```
Oh, one more thing: the indentation is off here, which is causing Checkstyle to fail the build:
```suggestion
                    log.error("Failed to sync configs for topic {} on cluster {} with IncrementalAlterConfigs API", k.name(), sourceAndTarget.target(), e);
                    context.raiseError(new ConnectException("use.incremental.alter.configs was set to \"required\", but the target cluster '"
                            + sourceAndTarget.target() + "' is not compatible with IncrementalAlterConfigs API", e));
```

Can you try to make sure that the project builds before pushing? `./gradlew :connect:mirror:build` should do the trick, or if you want to skip integration tests (since those can take a while) you can do `./gradlew :connect:mirror:build :connect:mirror:unitTest -x test`.

tinaselenge committed:
Address remaining comments

tinaselenge committed:
Merge branch 'trunk' into KAFKA-14420

tinaselenge commented:
@C0urante I have been meaning to get back to it and address the remaining comments. I think I have now addressed everything. Thank you so much for reviewing the PR again. Please let me know if I still missed something or have further comments.

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -514,6 +539,43 @@ private void updateTopicConfigs(Map<String, Config> topicConfigs) {
+            configOps.put(configResource, ops);
+        }
+        log.trace("Syncing configs for {} topics.", configOps.size());
+        targetAdminClient.incrementalAlterConfigs(configOps).values().forEach((k, v) -> v.whenComplete((x, e) -> {
```
Because we're using [AlterConfigsResult::values](https://kafka.apache.org/34/javadoc/org/apache/kafka/clients/admin/AlterConfigsResult.html#values()) instead of [AlterConfigsResult::all](https://kafka.apache.org/34/javadoc/org/apache/kafka/clients/admin/AlterConfigsResult.html#all()) here, the body after this line (starting with `if (e != null)` may be executed multiple times for a single invocation of `MirrorSourceConnector::incrementalAlterConfigs`.

Since we use the topic name in our error messages, it seems acceptable to use this specific API from the results class. But we should probably avoid logging the warning messages about falling back on the deprecated API or the error-level message about failing to sync the configs for the topic more than once.

Can we only log those messages once, and if there are any more failures, log the usual warning "Could not alter configurations of topic &lt;topic&gt;" message instead?

C0urante commented:
@tinaselenge I owe you a big apology!

I think the API you had for the `ConfigPropertyFilter` interface was correct the first time around.

The double-check logic is necessary because we need to distinguish between three possible cases when incrementally altering topic configs:

0. A property with a default value on the source topic should be synced to the target topic (using a `SET` operation in the config alter request)
0. A property with a default value on the source topic should use the default value on the target topic (which is accomplished via a `DELETE` operation in the config alter request)
0. A property with a default value on the source topic should not be altered on the target topic at all

If we only rely on `ConfigPropertyFilter::shouldReplicateSourceDefault`, that doesn't give us enough information to choose between those three options (since the return value of that method can only have two different values).

Instead, we can and should use both methods to decide how to act (for some property that has a default value on the source cluster):

0. If both `ConfigPropertyFilter::shouldReplicateConfigProperty` and `ConfigPropertyFilter::shouldReplicateSourceDefault` return `true`, then we sync the property to the target cluster using `SET`
0. If `ConfigPropertyFilter::shouldReplicateConfigProperty` returns `true` but `ConfigPropertyFilter::shouldReplicateSourceDefault` returns `false`, then we sync the property to the target cluster using `DELETE`
0. If `ConfigPropertyFilter::shouldReplicateConfigProperty` returns `false`, we do not sync the property at all

So I've come to realize that the way you'd originally implemented this is actually the only correct way to satisfy the design outlined in the KIP.

I'm very sorry for wasting your time on this wild goose chase, please accept my apologies!

tinaselenge committed:
revert back to double check logic in targetConfig().

tinaselenge commented:
@C0urante No worries at all! I should have clarified it better. I think both logics work but perhaps the double logic in the targetConfig() is better since we can't rely solely on shouldReplicateSourceDefault as you said. I reverted that part but also added some comments explaining the conditions that we are filtering against, hopefully that makes sense.

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/ConfigPropertyFilter.java:
```diff
@@ -25,8 +25,20 @@
+    /**
+     * Specifies whether to replicate the given topic configuration.
+     * This does not handle replication of configurations with default values.
+     * The configurations with default values are controlled by {@link #shouldReplicateSourceDefault(String)}.
```
This part should be updated now. Maybe something like this?

```suggestion
     * Note that if a property has a default value on the source cluster,
     * {@link #shouldReplicateSourceDefault(String)} will also be called to
     * determine how that property should be synced.
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/ConfigPropertyFilter.java:
```diff
@@ -25,8 +25,20 @@
     boolean shouldReplicateConfigProperty(String prop);
 
+    /**
+     * Specifies whether to replicate the given topic configuration that has a default value
```
I think we can give more detail on the exact behavior for this API:

```suggestion
     * Specifies how to replicate the given topic configuration property
     * that has a default value on the source cluster. Only invoked for properties
     * that {@link #shouldReplicateConfigProperty(String)} has returned
     * {@code true} for.
     *
     * @return {@code true} if the default value from the source topic should be synced
     * to the target topic, and {@code false} if the default value for the target topic
     * should be used instead
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/DefaultConfigPropertyFilter.java:
```diff
@@ -30,21 +30,29 @@ public class DefaultConfigPropertyFilter implements ConfigPropertyFilter {
-                                                                + "that should not be replicated.";
+                                                                + "that should not be replicated."
+                                                                + "This setting also applies to the properties with default values"
+                                                                + "if use.defaults.from is set to 'source'.";
```
I think we should remove this part since this property now affects the behavior for all properties, regardless of the value of `use.defaults.from`.
```suggestion
                                                                + "that should not be replicated.";
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -538,9 +600,13 @@ Map<String, Config> describeTopicConfigs(Set<String> topics)
+    Config targetConfig(Config sourceConfig, boolean incremental) {
+// If using incrementalAlterConfigs API, sync the default property with either SET or DELETE action determined by ConfigPropertyFilter::shouldReplicateSourceDefault later.
+// If not using incrementalAlterConfigs API, sync the default property if ConfigPropertyFilter::shouldReplicateSourceDefault returns true.
+// If ConfigPropertyFilter::shouldReplicateConfigProperty returns false, we do not sync the property at all.
```
Nit: can we align this with the next line of the method body?
```suggestion
        // If using incrementalAlterConfigs API, sync the default property with either SET or DELETE action determined by ConfigPropertyFilter::shouldReplicateSourceDefault later.
        // If not using incrementalAlterConfigs API, sync the default property if ConfigPropertyFilter::shouldReplicateSourceDefault returns true.
        // If ConfigPropertyFilter::shouldReplicateConfigProperty returns false, we do not sync the property at all.
```

C0urante commented on connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:
```diff
@@ -538,9 +600,13 @@ Map<String, Config> describeTopicConfigs(Set<String> topics)
+// If ConfigPropertyFilter::shouldReplicateConfigProperty returns false, we do not sync the property at all.
         List<ConfigEntry> entries = sourceConfig.entries().stream()
-            .filter(x -> !x.isDefault() && !x.isReadOnly() && !x.isSensitive())
+            .filter(x -> incremental || x.isDefault() && shouldReplicateSourceDefault(x.name()) || !x.isDefault())
```
Nit: a pair of parentheses here, though not strictly necessary, could help illustrate the behavior of this boolean expression a bit more clearly:
```suggestion
            .filter(x -> incremental || (x.isDefault() && shouldReplicateSourceDefault(x.name())) || !x.isDefault())
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -209,15 +229,62 @@ public void testConfigPropertyFiltering() {
+        assertTrue(targetConfig.entries().stream()
+                .anyMatch(x -> x.name().equals("name-1")), "should replicate properties");
+        assertTrue(targetConfig.entries().stream()
+                .anyMatch(x -> x.name().equals("name-2")), "should not replicate default properties");
```
```suggestion
                .anyMatch(x -> x.name().equals("name-2")), "should include default properties");
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +328,114 @@ public void testNewTopicConfigs() throws Exception {
+        props.put(MirrorSourceConfig.USE_INCREMENTAL_ALTER_CONFIGS, MirrorSourceConfig.REQUEST_INCREMENTAL_ALTER_CONFIGS);
+        MirrorSourceConfig connectorConfig = new MirrorSourceConfig(props);
+
+        MockAdminClient admin = spy(new MockAdminClient());
```
We don't actually have to use the `MockAdminClient` class at all and can just mock the `Admin` interface for this test:
```suggestion
        Admin admin = mock(Admin.class);
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +328,114 @@ public void testNewTopicConfigs() throws Exception {
+        props.put(MirrorSourceConfig.USE_INCREMENTAL_ALTER_CONFIGS, MirrorSourceConfig.REQUIRE_INCREMENTAL_ALTER_CONFIGS);
+        MirrorSourceConfig connectorConfig = new MirrorSourceConfig(props);
+
+        MockAdminClient admin = spy(new MockAdminClient());
```
(Same thought RE just mocking the `Admin` interface):
```suggestion
        Admin admin = mock(Admin.class);
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +328,114 @@ public void testNewTopicConfigs() throws Exception {
+            ops.add(new AlterConfigOp(entryWithNonDefaultValue, AlterConfigOp.OpType.SET));
+            ops.add(new AlterConfigOp(entryWithDefaultValue, AlterConfigOp.OpType.DELETE));
+
+            assertEquals(configOps.get(configResource), ops);
```
Nit: the ordering for this `assertEquals` method is expected first, then actual (this matters because the messages generated for assertion failures can be confusing otherwise):
```suggestion
            assertEquals(ops, configOps.get(configResource));
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +328,114 @@ public void testNewTopicConfigs() throws Exception {
+        props.put(MirrorSourceConfig.USE_INCREMENTAL_ALTER_CONFIGS, MirrorSourceConfig.REQUIRE_INCREMENTAL_ALTER_CONFIGS);
+        MirrorSourceConfig connectorConfig = new MirrorSourceConfig(props);
+
+        MockAdminClient admin = spy(new MockAdminClient());
```
```suggestion
        Admin admin = mock(Admin.class);
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +328,114 @@ public void testNewTopicConfigs() throws Exception {
+        Config config = new Config(entries);
+        doReturn(Collections.singletonMap(topic, config)).when(connector).describeTopicConfigs(any());
+        doReturn(alterConfigsResult(new ConfigResource(ConfigResource.Type.TOPIC, topic), new UnsupportedVersionException("Unsupported API"))).when(admin).incrementalAlterConfigs(any());
+        doNothing().when(connector).deprecatedAlterConfigs(any());
```
We no longer need this part:
```suggestion
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceConnectorTest.java:
```diff
@@ -261,6 +328,114 @@ public void testNewTopicConfigs() throws Exception {
+        doNothing().when(connector).deprecatedAlterConfigs(any());
+        connector.syncTopicConfigs();
+        Map<String, Config> topicConfigs = Collections.singletonMap("source." + topic, config);
+        verify(connector).deprecatedAlterConfigs(topicConfigs);
```
Can we also verify that we never tried to use the incremental API?
```suggestion
        verify(connector).deprecatedAlterConfigs(topicConfigs);
        verify(connector, never()).incrementalAlterConfigs(any());
```

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:
```diff
@@ -655,6 +657,135 @@ public void testRestartReplication() throws InterruptedException {
+            assertEquals("1000", backupConfig,
+                    "`retention.bytes` should be synced with default value!");
+
+            // delete the previously altered configuration of the source topic
+            ConfigResource configResource = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+            Collection<AlterConfigOp> ops = new ArrayList<>();
+            ops.add(new AlterConfigOp(new ConfigEntry("retention.bytes", "1000"), AlterConfigOp.OpType.DELETE));
+            Map<ConfigResource, Collection<AlterConfigOp>> configOps = Collections.singletonMap(configResource, ops);
+            primary.kafka().incrementalAlterConfigs(configOps);
```
Nit: any reason this has to take place inside the call to `waitForCondition`? Would it make more sense to move this part to start at line 778?

C0urante commented on connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:
```diff
@@ -655,6 +657,135 @@ public void testRestartReplication() throws InterruptedException {
+            assertEquals(primaryConfig, backupConfig,
+                    "`retention.bytes` should be the same, because it isn't in exclude filter! ");
+            return true;
+        }, 3000, "Topic configurations were not synced");
```
We should use a much higher timeout here (30 or 60 seconds). Our CI machines are notoriously slow and this may lead to flakiness as-is.

C0urante reviewed (commented):
Thanks @tinaselenge! This is looking great, almost ready to merge.

tinaselenge committed:
Merge branch 'trunk' into KAFKA-14420

tinaselenge committed:
Address the final comments

C0urante reviewed (approved):
LGTM, thanks @tinaselenge!

(Sorry for the delay; took last week off 🌴)

ijuma commented:
It looks like we started getting a _lot_ of test failures after this PR:

https://ci-builds.apache.org/job/Kafka/job/kafka/job/trunk/1746/#showFailuresLink

I see a lot of the same failures in this PR build:

https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13373/16/#showFailuresLink

@C0urante did you triage these failures before merging?

gharris1727 commented:
Ticket for the failures introduced here: https://issues.apache.org/jira/browse/KAFKA-14905

Some cursory investigation reveals that these tests hook into `alterConfigs` via `FakeForwardingAdminWithLocalMetadata` but don't hook the corresponding `incrementalAlterConfigs`. However, I cannot confirm this as the tests are consistently passing locally with or without that fix. I'll open a tentative fix PR and see if the test then passes on Jenkins.

ijuma commented:
Thanks for helping with this @gharris1727.
